src/services/application.js

```diff
@@ -18,6 +38,160 @@ class CompanyApplicationService {
+            )]
+                .map(async (application) => ({
+                    ...application.toObject(),
+                    state: (await CompanyApplication.findById(application._id).exec()).state,
+                }))),
+        };
+    }
+
+    buildFiltersQuery({
+        companyName,
+        submissionDate,
+    }) {
+        const filterQueries = [];
+
+        if (companyName) {
+            filterQueries.push({
+                // This allows for partial text matching.
+                // If only full-text is needed, use $text query instead, which uses
+                // a text index and is more performant
+                companyName: new RegExp(companyName, "gi"),
+            });
+        }
+
+        if (submissionDate) {
+
+            const filter = { $and: [] };
+
+            if (submissionDate.to) {
+                filter["$and"].push({ submittedAt: { "$lte": submissionDate.to } });
+            }
+            if (submissionDate.from) {
+                filter["$and"].push({ submittedAt: { "$gte": submissionDate.from } });
+            }
+            filterQueries.push(filter);
+        }
+
+        return filterQueries;
+    }
+
+    /**
+     * @param {*} limit - Number of documents to return
+     * @param {*} offset - where to start the query (pagination - how many documents to skip, NOT how many pages to skip)
+     * @param {*} filters - object with optional properties: state, submissionDate, companyName
+     * {
+     *      companyName: String
+     *      submissionDate: { - only one of the properties is necessary, but you can use both to define interval
+     *          from: Date
+     *          to: Date
+     *      }
+     *      state: String | Array - String for exact match, Array to provide set of options
+     * }
+     * @returns {applications, docCount}
+     */
+    async find(filters, limit, offset, sortingOptions) {
+
+        const { state: stateFilter, ...queryFilters } = { ...filters };
+
+        if (!filters || Object.keys(filters).length === 0) return this.findAll(limit, offset, sortingOptions);
+
+        const docCount = await CompanyApplication.estimatedDocumentCount();
+
+        // Using .skip().limit() can be problematic if we get big data,
+        // Once problems appear, consider using .cursor API
+
+        return {
+            docCount,
+            applications: (await Promise.all(
+                (await CompanyApplication.find(
+                    Object.keys(queryFilters).length ? {
+                        $and: this.buildFiltersQuery(queryFilters),
+                    } : {})
+                    .sort(sortingOptions || { submittedAt: "desc" })
+                    .skip(offset)
+                    .limit(limit)
+                    .exec()
+                )
+                    .map(async (application) => ({
+                        ...application.toObject(),
+                        state: (await CompanyApplication.findById(application._id).exec()).state,
```

Comment:
The state can be determined by the other fields in the application, this will save a lot of unnecessary queries.

Reply:
Just found this is a thing: https://mongoosejs.com/docs/api.html#document_Document-toObject

Also took the chance and removed passwords from the returned objects so no more: `delete password; return retVal` ugliness